### ISSUE TITLE
EIP7702 questions

### ISSUE THREAD

**Issue by winsvega** [ethereum/execution-spec-tests#797]
Delegation tests questions:

1)
so if we send tx from sender to sender it will change the code of the sender
to delegation_designation. this will make sender not EOA or we must redefine the meaning of EOA
because in all of the tests we explicitly removed all the code from tx senders. 
and now it will be possible to broadcast transactions from accounts that "have code" event if its delegation code.

2) EIP spec 
assert auth.y_parity < 2**256
assert auth.r < 2**256
assert auth.s < 2**256
actually we limit signature in Yellow paper appendinx F
0 < s < secp256k1n ÷ 2 + 1

3)
Verify the chain id is either 0 or the chain's current ID.
hm, I think it must enforce chain id to be of current chain id. otherwise you can steal this auth from L2 chain?

4)
Set the code of authority to be 0xef0100 || address. This is a delegation designation.
so the delegations don't stack each time it overrides? 

5)
If any of the above steps fail, immediately stop processing that tuple and continue to the next tuple in the list.
potential to spam invalid tuples.

**Comment by marioevz** [comment]
>     1. TESTS
>        so if we send tx from sender to sender it will change the code of the sender
>        to delegation_designation. this will make sender not EOA or we must redefine the meaning of EOA
>        because in all of the tests we explicitly removed all the code from tx senders.
>        and now it will be possible to broadcast transactions from accounts that "have code" event if its delegation code.
> 
> ah I see now need to modify 3607. this need blockchain testing.

See https://github.com/ethereum/execution-spec-tests/blob/589ab4ef1daaf97dfa60856ca0bed83891a684f9/tests/prague/eip7702_set_code_tx/test_set_code_txs.py#L2446

> 
>     2. EIP spec
>        assert auth.y_parity < 2**256
>        assert auth.r < 2**256
>        assert auth.s < 2**256
>        actually we limit signature in Yellow paper appendinx F
>        0 < s < secp256k1n ÷ 2 + 1

See https://github.com/ethereum/execution-spec-tests/blob/589ab4ef1daaf97dfa60856ca0bed83891a684f9/tests/prague/eip7702_set_code_tx/test_set_code_txs.py#L1947

But this needs to be expanded because it's going to probably change for devnet-4.

>     3. EIP spec
>        Verify the chain id is either 0 or the chain's current ID.
>        hm, I think it must enforce chain id to be of current chain id. otherwise you can steal this auth from L2 chain?

There are a couple of things here:
- The auth from another L2 will only let an attacker point your account to the same address in every chain, and I think it would be very hard for a malicious contract to land on the same address in a different EVM compatible chain. The only way I could see a potential attack is that you sign a set-code to a proxy, and this proxy then gets compromised on a different chain.
- Your account must have the same nonce as the authorization you are trying to mimic for it to go through.

>     4. EIP spec
>        Set the code of authority to be 0xef0100 || address. This is a delegation designation.
>        so the delegations don't stack each time it overrides?

Yes exactly, that's how you can "cancel" a delegation designation, you send another type-4 transaction to point to, maybe 0x0000000000000000000000000000000000000000.

One missing test could be to try to deploy a contract with the delegation designation, which is already disallowed by https://eips.ethereum.org/EIPS/eip-3541, but we need to verify that this is still not possible.

>     5. for the tests
>        check that EXTCODESIZE, EXTCODECOPY, EXTCODEHASH, CALL, CALLCODE, STATICCALL, DELEGATECALL  are redirected by the delegation pointer
See https://github.com/ethereum/execution-spec-tests/blob/589ab4ef1daaf97dfa60856ca0bed83891a684f9/tests/prague/eip7702_set_code_tx/test_set_code_txs.py#L896, https://github.com/ethereum/execution-spec-tests/blob/589ab4ef1daaf97dfa60856ca0bed83891a684f9/tests/prague/eip7702_set_code_tx/test_set_code_txs.py#L990, https://github.com/ethereum/execution-spec-tests/blob/589ab4ef1daaf97dfa60856ca0bed83891a684f9/tests/prague/eip7702_set_code_tx/test_set_code_txs.py#L1142, https://github.com/ethereum/execution-spec-tests/blob/589ab4ef1daaf97dfa60856ca0bed83891a684f9/tests/prague/eip7702_set_code_tx/test_set_code_txs.py#L1207, https://github.com/ethereum/execution-spec-tests/blob/589ab4ef1daaf97dfa60856ca0bed83891a684f9/tests/prague/eip7702_set_code_tx/test_set_code_txs.py#L595

>     6. EIP spec
>        In case a delegation designator points to another designator, creating a potential chain or loop of designators, clients must retrieve only the first code and then stop following the designator chain.
>        IMHO better OOG here. each time you introduce skip or continue logic its a potential bug

Well this is what happens in practice because the first opcode encountered when called into such scenario is `EF`, which is an invalid opcode, so you end up OOG'd.

>     7. EIP spec
>        The transaction sender will pay for all authorization tuples, regardless of validity or duplication.
>        again better to just verify this on transaction validation. to avoid potential spam. even if you add some gasprice.

This was discussed in ACD this week, and I think a very good point towards simply charging is to not complicate transaction validation verification and cause some potential DoS. Take for example the case where a type-4 transaction is received with a ton of authorizations in it, if authorization validity is necessary for the transaction to be valid, you need to go through each one of the authorizations, make one query to the state (to check if the authority has the correct nonce for example) per authorization, and if just the last authorization is invalid, you basically wasted a ton of effort to validate a transaction which cannot be included, and hence cannot be charged for. This would be a big DoS attack vector.

>     8. EIP spec
>        So if we can basically have our controlled address (EOA) to point to some external contarct.
>        Could it happen that I could make my EOA address to act as some defi proxy address, and if in some protocoll there is a call back or transfer back it can potentially hack the defi protocol?
>        like in a situation where   defiA  calls defiB  and defiB looks: "ok, the defiA  is my buddy, there for I transfer some funds to it"
>        And we make  EOA  to delegate to defiA,   then we send transaction from EOA to defiB, defiB looks: "hm, looks like I called by my buddy..."

So basically use a delegated EOA as a proxy. I think this would be a massive red-flag for a protocol, and would visible on-chain that this proxy has a delegation authorization, and therefore has a single external key to manipulate the proxy at any point.

**Comment by spencer-tb** [comment]
Closing. I believe we have tests for all cases questioned and discussed here.
